### PR TITLE
feat(genesis): Add flag for importing evm state to genesis.

### PR DESCRIPTION
**Pull request by vldm**
In order to import genesis argument `--evm-root` with last known evm state root, should be added to solana-genesis.
Also, solana-genesis expect `evm_genesis.bin` in ledger path. Currently this file contain evm state in json format.
evm_genesis.bin could be exported from parity blockchain, using `export state` command.

### CODE REVIEW

File: run.sh [hunk 1]
```diff
@@ -38,75 +38,77 @@ SOLANA_RUN_SH_CLUSTER_TYPE=${SOLANA_RUN_SH_CLUSTER_TYPE:-development}
...
-  # shellcheck disable=SC2086
-  solana-genesis \
+    ./fetch-spl.sh
+    if [[ -r spl-genesis-args.sh ]]; then
+        SPL_GENESIS_ARGS=$(cat spl-genesis-args.sh)
+    fi
+    
+    # shellcheck disable=SC2086
+    solana-genesis \
     --hashes-per-tick sleep \
```
Comment (hrls) [inline run.sh:73]: extra indent

```diff
diff --git a/run.sh b/run.sh
index e8b4597e0..3fbd6e820 100755
--- a/run.sh
+++ b/run.sh
@@ -67,21 +67,21 @@ else
     if [[ -r spl-genesis-args.sh ]]; then
         SPL_GENESIS_ARGS=$(cat spl-genesis-args.sh)
     fi
-    
+
     # shellcheck disable=SC2086
     solana-genesis \
-    --hashes-per-tick sleep \
-    --faucet-lamports 500000000000000000 \
-    --bootstrap-validator \
-    "$dataDir"/validator-identity.json \
-    "$dataDir"/validator-vote-account.json \
-    "$dataDir"/validator-stake-account.json \
-    --ledger "$ledgerDir" \
-    --cluster-type "$SOLANA_RUN_SH_CLUSTER_TYPE" \
-    $SPL_GENESIS_ARGS \
-    --max-genesis-archive-unpacked-size=300000000 \
-    $SOLANA_RUN_SH_GENESIS_ARGS
-    # --evm-root="0x7b343e0165c8f354ac7b1e7e7889389f42927ccb9d0330b3036fb749e12795ba" \
+        --hashes-per-tick sleep \
+        --faucet-lamports 500000000000000000 \
+        --bootstrap-validator \
+        "$dataDir"/validator-identity.json \
+        "$dataDir"/validator-vote-account.json \
+        "$dataDir"/validator-stake-account.json \
+        --ledger "$ledgerDir" \
+        --cluster-type "$SOLANA_RUN_SH_CLUSTER_TYPE" \
+        $SPL_GENESIS_ARGS \
+        --max-genesis-archive-unpacked-size=300000000 \
+        $SOLANA_RUN_SH_GENESIS_ARGS
+    # --evm-root="0x7b343e0165c8f354ac7b1e7e7889389f42927ccb9d0330b3036fb749e12795ba"
 fi
 
 abort() {
```

File: evm-utils/evm-rpc/src/serialize.rs [hunk 2]
```diff
@@ -5,8 +5,7 @@ use std::str::FromStr;
 use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
 
 use primitive_types::{H160, H256, H512, U128, U256, U512};
-
-#[derive(Debug, Default, Hash, Clone, PartialEq, Eq)]
+#[derive(Debug, Default, Ord, PartialOrd, Hash, Clone, Eq, PartialEq)]
```
Comment (hrls) [inline evm-utils/evm-rpc/src/serialize.rs]: `Copy` too

File: evm-utils/evm-state/src/layered_backend.rs [hunk 3]
```diff
@@ -313,16 +313,29 @@ impl EvmState {
...
-            fs::create_dir(path)?;
+        if evm_state.is_dir() && evm_state.exists() {
+            warn!("deleting existing state {}", evm_state.display());
+            fs::remove_dir_all(&evm_state)?;
+            fs::create_dir(&evm_state)?;
         }
 
-        Self::load_from(path, Slot::default(), empty_trie_hash())
+        let root_hash = if let Some(root_hash) = root_hash {
+            let evm_genesis = path.as_ref().to_path_buf().join("evm-state-genesis");
```
Comment (hrls) [inline evm-utils/evm-state/src/layered_backend.rs]: `Path::join`, so `Path::to_path_buf` is unnecessary.

File: evm-utils/evm-state/src/layered_backend.rs [hunk 4]
```diff
@@ -313,16 +313,29 @@ impl EvmState {
...
-            fs::remove_dir_all(path)?;
-            fs::create_dir(path)?;
+        if evm_state.is_dir() && evm_state.exists() {
+            warn!("deleting existing state {}", evm_state.display());
+            fs::remove_dir_all(&evm_state)?;
+            fs::create_dir(&evm_state)?;
         }
 
-        Self::load_from(path, Slot::default(), empty_trie_hash())
+        let root_hash = if let Some(root_hash) = root_hash {
```
Comment (hrls) [inline evm-utils/evm-state/src/layered_backend.rs]: Why this logic is here and not in somewhere near genesis related code?
I prefer move this out of `EvmState::new` and reuse existing `Storage::restore_from` and then `EvmState::load_from` in some task specific new function like `EvmState::restore_and_load/load_genesis(backup_dir, target_dir, root)`.

File: evm-utils/evm-state/src/layered_backend.rs [hunk 5]
```diff
@@ -313,16 +313,29 @@ impl EvmState {
 }
 
 impl EvmState {
-    pub fn new<P: AsRef<Path>>(path: P) -> Result<Self, anyhow::Error> {
-        let path = path.as_ref();
+    pub fn new<P: AsRef<Path>>(path: P, root_hash: Option<H256>) -> Result<Self, anyhow::Error> {
```
Comment (hrls) [inline evm-utils/evm-state/src/layered_backend.rs]: Root hash is always None or default empty for new EvmState.

File: genesis/src/main.rs [hunk 6]
```diff
@@ -511,6 +522,22 @@ fn main() -> Result<(), Box<dyn error::Error>> {
...
+    if cfg!(feature = "with_evm") {
+        let root = value_t!(matches, "evm-root", String);
+        match root {
+            Ok(root) => {
+                let root_hash = evm_rpc::Hex::<evm_state::H256>::from_hex(&root).unwrap();
+                genesis_config.set_evm_state_file(root_hash.0)
+            }
+            Err(e) => {
+                eprintln!(
+                    "EVM root was not found but genesis was compilet with `with_evm` feature {}",
```
Comment (hrls) [inline genesis/src/main.rs]: typo fix
```suggestion
                    "EVM root was not found but genesis was compiled with `with_evm` feature {}",
```

File: ledger/src/bank_forks_utils.rs [hunk 7]
```diff
@@ -32,7 +32,7 @@ fn to_loadresult(
 pub fn load(
     genesis_config: &GenesisConfig,
     blockstore: &Blockstore,
-    evm_state_path: PathBuf,
+    evm_state_root_path: PathBuf,
```
Comment (hrls) [inline ledger/src/bank_forks_utils.rs]: Looks like a path to EvmState.root

File: sdk/src/genesis_config.rs [hunk 8]
```diff
@@ -198,7 +207,53 @@ impl GenesisConfig {
...
+                cluster_type => {
+                    return Err(std::io::Error::new(
+                        std::io::ErrorKind::Other,
+                        format!("Trying to generate genesis for cluster = {:?} without evm state root hash provided.", cluster_type),
+                    ))
+                }
+            }
+        };
+        let mut evm_genesis_path = ledger_path.to_path_buf();
+        evm_genesis_path.push(EVM_STATE_FILE);
```
Comment (hrls) [inline sdk/src/genesis_config.rs]:
```rust
ledger_path.join(EVM_STATE_FILE)
```

Comment (hrls) [inline sdk/src/genesis_config.rs]: `EVM_STATE_GENESIS_FILE` or more convenient name

File: sdk/src/genesis_config.rs [hunk 9]
```diff
@@ -198,7 +207,53 @@ impl GenesisConfig {
...
+        evm_backup.push("evm-state-genesis");
+
+        let tmp_backup = evm_state
+            .kvs
+            .backup()
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, format!("{}.", e)))?;
+        std::fs::rename(tmp_backup, evm_backup)
+    }
+
+    pub fn set_evm_state_file(&mut self, root_hash: H256) {
```
Comment (hrls) [inline sdk/src/genesis_config.rs]: Wrong name

File: run.sh (hunk 1 above)
Comment (vldm) [reply run.sh:73]: My formatter don't like this.

**Review by hrls — APPROVED** [review]

**Pull request merged**